### PR TITLE
Fix bug: fourslash test does not recognize tsconfig.json.

### PR DESCRIPTION
**Pull request by hoanhtien**
Adding `// @Filename: tsconfig.json` to a fourslash test makes no effect. The reason is we are using `ts.readConfigFile()` which reads the config file from disk. Change to `this.host.readFile()` instead.

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
Fourslash already supports directive comments (e.g. `//@module: commonjs`). What's the intent of this?

**Comment by danquirk** [comment]
Along the lines of what Ryan mentions, if we did this we need:
- fourslash to give an error when you mix directives and tsconfig
- more tests on the various ways in which this works/would work (particularly for more complex file arrangements) to ensure that future tests are doing what we expect and people don't start writing lots of tsconfig based tests that give a false sense of security (ex fourslash does its own weird file resolution that doesn't match the compiler itself)

**Comment by hoanhtien** [comment]
@RyanCavanaugh The intent here is to some kind of test to verify that `tsconfig.json` is found and parsed correctly. We don't have this kind of test right now and moving forward there will be more work requiring it e.g. I am planning to expose `tsconfig.json` errors by adding it into `tsserver` response for `open` and `projectInfo` commands.
@danquirk Those are good points. If you guys think fourslash is a decent way to test `tsconfig.json`, I will move forward with improving this pull request.

**Comment by mhegazy** [comment]
@RyanCavanaugh we have no way to test tsconfig.json LS scenarios. this adds that.

File: src/harness/fourslash.ts [hunk 1]
```diff
@@ -2485,6 +2496,34 @@ module FourSlash {
         };
     }
 
+    function containTSConfigJson(files: FourSlashFile[]): boolean {
+        for (let i = 0; i < files.length; ++i) {
```
Comment (mhegazy) [inline src/harness/fourslash.ts]: can be reduced to:
```ts
return forEach(files, f => f.fileOptions['Filename'] === 'tsconfig.json');
```

File: src/harness/fourslash.ts [hunk 2]
```diff
@@ -2485,6 +2496,34 @@ module FourSlash {
...
+    }
+
+    function getNonFileNameOptionInFileList(files: FourSlashFile[]): string {
+        for (let i = 0; i < files.length; ++i) {
+            let option = getNonFileNameOptionInObject(files[i].fileOptions);
+            if (option !== null) {
+                return option;
+            }
+        }
+        return null;
```
Comment (mhegazy) [inline src/harness/fourslash.ts]: do not use `null` use `undefined`.

Comment (mhegazy) [inline src/harness/fourslash.ts]:
```ts
return ts.forEach(files, f => getNonFileNameOptionInObject(f.fileOptions));
```

File: src/harness/fourslash.ts [hunk 3]
```diff
@@ -2476,6 +2476,17 @@ module FourSlash {
             }
         }
 
+        // @Filename is the only directive that can be used in a test that contains tsconfig.json file.
+        if (containTSConfigJson(files)) {
+            let directive = getNonFileNameOptionInFileList(files);
+            if (directive == null) {
```
Comment (mhegazy) [inline src/harness/fourslash.ts]: do not use `null` either `if (directive)` or `if (directive ==== undefined)`

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**